### ISSUE TITLE
Plugin works but slows my site right down

### ISSUE THREAD

**Issue by sdesign75** [henrywright/buddypress-identicons#1]
I've tried your plugin quite a few times and it works beautifully except my page generation times increase from 1 second to over 8 seconds when I have pixicons enabled. I've analysed the site and it shows me that your plugin is causing 65% of the total plugin resources (and I have over 40 plugins installed so thats pretty significant). If I change back to Retro everything is fine again. I'm using WP 4.4 and Buddypress.

Thanks,

Alex

**Comment by henrywright** [comment]
Thanks for the report @sdesign75 

The heavy load is likely due to image generation. Image generation happens once only for each of your website's members. After the initial image generation, images are stored in the `uploads/` directory which means far less resources are used. 

If you rerun the tool for a second time on the same page with the same identicons displayed, you should notice far less resources are used.

**Comment by sdesign75** [comment]
No that's not the case. I just checked again. When the plugin is activated a page takes 8 secs to generate a page, and this is the same every time I refresh the same page. I've when through my members directory and then checked by ftp and all the images have been created and cached but they are just just making the page load much longer. I also have a caching plugin and checked the page generation time with that and it is 2 secs without your plugin and 8 secs with it. When I keep your plugin active but change the gravatar to Retro everything is fine with quick page speed.

**Comment by henrywright** [comment]
8s page load is horrible, hopefully we can resolve this.

Right now I'm thinking image size can't be the issue because by default each image is a tiny ~500 bytes. If you really wanted to you could run an imagemin script recursively to the `uploads/` directory but I don't think that'll give huge gains in this case.

Can you upload one of the generated images here so I can take a look at the file?

Also, image lookup shouldn't take too much effort either because all images are local. Unlike Gravatar, there's calls being made to a third-party service.

Perhaps a plugin conflict is happening. If you disable everything else and run TwentyFifteen, does the page still take 8s to load?

Which version of the plugin are you using?

**Comment by sdesign75** [comment]
Yes, the image is pretty small. I've attached one for you. I don't think it's a plugin conflict because I your plugin was one of the first I installed when building the site and I uninstalled it straight away because of the effect on page speed. If you are getting 2s with Twenty Fifteen and no other plugins I'd say that was quite high too TBH. What page speed do you get without your plugin activated?

Thanks for looking into this!

Alex

![pixicon](https://cloud.githubusercontent.com/assets/16734744/12517491/d879acc4-c12a-11e5-84f4-ea7b8dc731b4.png)

**Comment by henrywright** [comment]
I tested about 6 times on localhost using [Chrome Dev Tools](https://developer.chrome.com/devtools) and took 2s as the average.

4.6s was the highest (I got that on my first try). The next 5 tests were all just over 1s.

Just done the same again with my plugin deactivated, the first try was 2.8 seconds and the rest were all just over 1s.

Can you deactivate all plugins aside from BuddyPress and BuddyPress Identicons and activate Twenty Fifteen and then let me know what load time you get?

**Comment by henrywright** [comment]
@sdesign75 

I did some more investigating and found that in order to check if a user has uploaded a Gravatar image, `identicons_grav_exists()` sends an extra request to Gravatar and waits for a response. This happens for each user on the page so could account for the delay you're seeing.

I've submitted patch [#6989](https://buddypress.trac.wordpress.org/ticket/6989) upstream which will allow Gravatar's `d` param to be filtered. This will avoid the need for extra communication with Gravatar.
